src/relay/backend/graph_runtime_codegen.cc

```diff
@@ -0,0 +1,761 @@
+      const auto& storage_id = dmlc::get<std::vector<int64_t>>(node->attrs_["storage_id"]);
+      const auto& dtype_vec = dmlc::get<std::vector<std::string>>(node->attrs_["dtype"]);
+
+      CHECK_EQ(node->num_outputs_, shape_vec.size());
+      num_entry += node->num_outputs_;
+
+      shapes.insert(shapes.end(), shape_vec.begin(), shape_vec.end());
+      dltypes.insert(dltypes.end(), dtype_vec.begin(), dtype_vec.end());
+      storage_ids.insert(storage_ids.end(), storage_id.begin(), storage_id.end());
+      if (node->attrs_.count("device_index")) {
+        const auto& dev_types = dmlc::get<std::vector<int64_t>>(node->attrs_["device_index"]);
+        device_types.insert(device_types.end(), dev_types.begin(), dev_types.end());
+      }
+      node_row_ptr.push_back(num_entry);
+    }
+    writer->BeginObject();
+    writer->WriteObjectKeyValue("nodes", nodes_);
+    writer->WriteObjectKeyValue("arg_nodes", arg_nodes);
+    writer->WriteObjectKeyValue("heads", heads_);
+    std::unordered_map<std::string, std::vector<dmlc::any>> attrs;
+    attrs["shape"].emplace_back(std::string("list_shape"));
+    attrs["shape"].emplace_back(shapes);
+    attrs["storage_id"].emplace_back(std::string("list_int"));
+    attrs["storage_id"].emplace_back(storage_ids);
+    if (device_types.size()) {
+      attrs["device_index"].emplace_back(std::string("list_int"));
+      attrs["device_index"].emplace_back(device_types);
+    }
+    attrs["dltype"].emplace_back(std::string("list_str"));
+    attrs["dltype"].emplace_back(dltypes);
+    writer->WriteObjectKeyValue("attrs", attrs);
+    writer->WriteObjectKeyValue("node_row_ptr", node_row_ptr);
+    writer->EndObject();
+  }
+
+  /*!
+   * \brief Get unique name for func
+   *
+   * \param name
+   * \return std::string
+   */
+  std::string _GetUniqueName(const std::string& name) {
+    if (!name_map_.count(name)) {
+      name_map_[name] = 1;
+      return name;
+    }
+    auto index = name_map_[name];
+    name_map_[name] += 1;
+    return _GetUniqueName(name + std::to_string(index));
+  }
+
+ protected:
+  /*! \brief nodes */
+  std::vector<GraphNodePtr> nodes_;
+  /*! \brief output of graph */
+  std::vector<GraphNodeRef> heads_;
+  /*! \brief mod */
+  runtime::Module* mod_;
+  /*! \brief variable map */
+  std::unordered_map<const Node*, std::vector<GraphNodeRef>> var_map_;
+  /*! \brief target device */
+  TargetsMap targets_;
+  /*! \brief params */
+  std::unordered_map<std::string, runtime::NDArray> params_;
+  /*! \brief plan memory of device result */
+  Map<Expr, Array<IntegerArray>> storage_device_map_;
+  /*! \brief lowered funcs */
+  std::unordered_map<std::string, std::unordered_set<LoweredFunc, NodeHash, NodeEqual>>
+      lowered_funcs_;
+  /*! \brief name map */
+  std::unordered_map<std::string, size_t> name_map_;
+  /*! \brief compile engine */
+  CompileEngine compile_engine_;
+};
+
+class GraphRuntimeCodegenModule : public runtime::ModuleNode {
```

Comment:
I think this makes sense for now, but in general I'm not a huge fan of reusing tvm::Module to implement this stringly typed function lookup map. Is there a better way to do this @tqchen ?

Reply:
It is a stateful function. I think Module is a convenient & clean way to do this.

Reply:
Yeah I understand, the string based dispatch of module isn't my favorite design we just did the same thing in the VTA simulator stuff.